generate_protos.py

```diff
@@ -0,0 +1,213 @@
+    parser.add_argument('--with_gateway', '-wg', default=False, action="store_true",
+                        help="Generate the grpc gateway files(experimental)")
+    parser.add_argument('--go_source_relative', default=False, action="store_true",
+                        help="Make go import paths 'source_relative - see https://github.com/golang/protobuf#parameters")
+    parser.add_argument('--validate_out', '-v', default=False, action="store_true", help="Specifies the input to protoc-gen-validate,"
+                                                                  " if not specified the plugin won't be invoked")
+
+    # Validation for constraints that aren't simply verifiable via argparse.
+
+    # Only go outputs can generate grpc-gateway.
+    args = parser.parse_args()
+    if args.language != "go" and args.with_gateway:
+        raise argparse.ArgumentError("Generating grpc-gateway is Go specific")
+
+    # The output directory is under `gen/pb<-LANGUAGE/_python>
+    # Python needs underscore for the output directory to be able to read it.
+    output_dir = OUTPUT_DIR + OUTPUT_DIR_PREFIX
+    if args.language == 'python':
+        output_dir += '_python'
+    else:
+        output_dir += '-' + args.language
+
+    # The output directory needs to exists for protoc to generate outputs.
+    os.makedirs(output_dir, exist_ok=True)
+
+    # subprocess.check_output uses an array for passing in arguments. The first arg is the
+    # binary name.
+    protoc_args = ['protoc']
+
+    # The list of includes will be reused if --with_gateway is specified.
+    # The first parameters to protoc are the includes required to resolve the protos.
+    includes =list(DEFAULT_INCLUDES)
+    # Append the extra includes.
+    for extra_include in args.includes:
+        includes.append("-I="+extra_include)
+
+    # Only one of file or directory can be specified.
+    if args.file:
+        # If a single file is specified as input to the script, then the current working directory is added
+        # to the includes prior to other includes.
+        includes = ["-I=."] + includes
+    else:
+        # For a directory, the other includes are specified prior to the directory with the protos.
+        includes.append("-I="+args.directory)
+
+    protoc_args.extend(includes)
+
+    # The next set of parameters are the language configurations.
+    if args.language == 'go':
+        go_parameters = "--go_out="
+        if args.go_source_relative:
+            go_parameters+=SOURCE_RELATIVE_PARAM
+        go_parameters+="plugins=grpc:"+output_dir
+        protoc_args.append(go_parameters)
+    elif args.language == "protodoc":
+        protoc_args += ["--plugin=protoc-gen-protodoc="+PROTODOC_PATH,
+        "--protodoc_out="+output_dir]
+    else:
+        protoc_args.append("--"+args.language+"_out="+output_dir)
+        if args.language != "java":
```

Comment:
I think java plugin is named a bit differently as `/usr/local/bin/protoc-gen-grpc-java`. We can either fix it here with a if/else branch, or create a softlink when building docker image.

Reply:
Do we plan to fix this?